feat: add documentation for recovery using `code`s (#995)
ory/docs: docs/kratos/self-service/flows/account-recovery-password-reset.mdx

---
id: account-recovery-password-reset

> title: Account recovery and password reset

Author: I feel like, here, it might be quite useful to have "Password reset" as well, as most people associate account recovery with password reset... But of course it's duplicated as well... Hm...

---

```mdx-code-block
import RenderFlow from "@site/src/theme/Code/RenderFlow"

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"

> ```

Author: Doesn't hurt either, does it? :)

And it helps with readability because it's highlighted as one block in my editor and thus separated from the actual text. So maybe we should start doing that in the other docs as well:

<img width="1095" alt="image" src="https://user-images.githubusercontent.com/5540255/199472381-aa038319-bb42-4bc5-9bba-e777a974855a.png">

> ```

Reviewer: Sure, no harm in that. There are lots of other documents that don't do that already in our docs set (even in this PR) that don't do that. I'm trying to introduce consistency to curb the chaos that sometimes governs our docs. For that reason, I'd like to see that wrapping removed, pls :)

> ```

Reviewer 2: > I don't think the import themselves must be wrapped - haven't seen this in any other docs.

They do need to be wrapped @tomekpapiernik . Please check the README of this repository ;) All code snippets are wrapped there. It's quite important to avoid problems in rendering.

https://github.com/ory/docs#import-markdown

> ```

Reviewer: Okay, so this was ignored in many cases. Gonna have to make an effort to fix that in the repo. Will update this PR.

       "type": "object",
       "properties": {
         "email": {
           "type": "string",
           "format": "email",
           "ory.sh/kratos": {
             "credentials": {
               "password": {
                 "identifier": true
               }
             },
+            "recovery": {
+              "via": "email"
+            }
           }
         }
       }
       "additionalProperties": false
     }
   }
 }
```

## Email templates

Ory Identities (Ory Kratos) provides a set of default email templates for the recovery flows.

:::important

In case a user tries to recover an unregistered account, an email is sent out to the email, informing them about the attempted
recovery.

This prevents account enumeration attacks as explained in this
[blog post by Troy Hunt](https://www.troyhunt.com/website-enumeration-insanity-how-our-personal-data-is-leaked/).

:::

To provide your own templates, see [Customizing E-Mail Templates](../../emails-sms/custom-email-templates).

## Updating credentials

After the user completes the recovery successfully they receive a privileged session and are redirected to the settings page to
update their credentials.

> Read more about the [privileged sessions](../../session-management/01_overview.mdx#privileged-sessions).

Reviewer: I'd move this information to the paragraph that opens this document. Would fit in well with the description of the general idea of this flow. If you're okay with that I'm happy to do it.

Author: You mean the introductory paragraph at the start? Feel free to do that, I am not exactly sure how that would work out.

## Invalidate other sessions

In some scenarios it can be useful to revoke all active sessions once an account has been recovered. This forces anyone with

> access to the account to re-authenticate.

Reviewer: To get that working in Ory Network you'd have to edit the config directly, there's no way to do that through the UI, right? We should explain that. 
Also, needs explanation that it's done through hooks (so you know that even before you look at the config snippet) and it's a good opportunity to our hooks docs.

Author: Good point! Done.

---
id: account-recovery-password-reset
title: Account recovery and password reset
---

import useBaseUrl from "@docusaurus/useBaseUrl"
import Mermaid from "@site/src/theme/Mermaid"
import ApiWarning from "@site/src/theme/ApiWarning"
import SelfServiceBrowserFlow from "@site/src/theme/SelfServiceBrowserFlow"
import SelfServiceSpaFlow from "@site/src/theme/SelfServiceSpaFlow"
import SelfServiceApiFlow from "@site/src/theme/SelfServiceApiFlow"

import CodeTabs from "@site/src/theme/Code/CodeTabs"
import {
  getFlowMethodLinkWithErrors,
  getFlowMethodLinkSuccess,
  getFlowMethodLinkInvalidChallenge,
  getFlowMethodLinkChallengeDone,

> } from "./code/recovery"

Author: Oh damn, you're right, forgot to clean up.

There were a lot of unused imports in there :D

import RenderFlow from "@site/src/theme/Code/RenderFlow"

import DomainWarning from "@site/src/theme/DomainWarning"
import AjaxWarning from "@site/src/theme/AjaxWarning"
import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"

Account Recovery allows registered users to regain access to their account if they lost their password or multifactor

> authentication (MFA) provider.

Author: Google Authenticator is just an implementation of the TOTP scheme. Google does provide the app, but AFAIK you can use it without having a Google account.

That being said, the Authenticator app _is_ the provider of the second factor, not Google.

But I like your solution better, anyway.

If a user loses the credentials to their account, they need to prove ownership of the account to recover access. This is typically
done via a recovery address such as an email address or phone number. The user can prove their identity by accessing a link or

> one-time code sent to a recovery address.

Author: Yes, sounds good.

> If a user lost the credentials to their account but still wants to access it, they need to prove ownership of the account. This is
> typically done via a recovery address, such as an email address or phone number. By having access to the inboxes of these
> addresses, the user proves to the system, that they are the same person, that initially registered the account.

Reviewer 3: ```suggestion
If a user loses the credentials to their account, they need to prove ownership of the account to recover access. This is typically done via a recovery address such as an email address or phone number. The user can prove their identity by accessing a link or one-time code sent to a recovery address.
```

Ory Identities (Ory Kratos) supports two recovery methods:

- Sending a one-time password (**preferred**)
- Sending a "magic link"

## Configuring self-service account recovery

<Tabs
  defaultValue="oc"
  values={[
    {label: 'The Ory Network', value: 'oc'},
    {label: 'Self-hosted Ory Kratos config', value: 'oss'}
  ]}>

>   <TabItem value="oc">

Reviewer: 1. Please wrap all MDX elements (such as <Tabs>) in ```mdx-code-block
2. The way you're doing the tab names is a bit too complicated I think, you can do it like this instead: 

```
   <Tabs groupId="react-or-preact">
     <TabItem value="react" label="React" default>
```

`groupId` allows you to set up the tabs so that all tabs are in sync
`default` allows you to decide which item is shown by default in the given tab group

Reviewer: https://github.com/ory/docs/blame/master/docs/getting-started/custom-ui-preact-ory-elements.mdx#L42

Author: Great point!! Thanks! This is way better.

> Account recovery can be configured in the **Customize** > **Account Recovery** screen of the
> [Ory Network Console](https://console.ory.sh/guides).

Reviewer: ```suggestion
To adjust account recovery options, go to **Authentication** → **Account Recovery** in the [Ory Console](https://console.ory.sh/guides)
```

  </TabItem>
  <TabItem value="oss">

You can configure which methods to use in the Ory Kratos config:

```yaml title="path/to/my/kratos/config.yml"
selfservice:
  methods:
    code: # this enables the one time password method
      enabled: true
      config:
        # Defines how long a recovery code is valid for (default 1h)
        lifespan: 15m

  flows:
    recovery:
      enabled: true

      # Defines how long a recovery flow (the UI interaction) is valid for (default 1h)
      lifespan: 15m
      use: code # Define which strategy to use, either `code` or `link`
```

Keep in mind, that both methods require sending out emails to your users, so an SMTP server is required. Read more
[here](../../emails-sms/01_sending-emails-smtp.mdx).

  </TabItem>
</Tabs>

## Account recovery addresses

> To provide the account recovery mechanism, Ory Identities (Ory Kratos) needs to know which address to send the recovery message
> to. In most cases, this is the address that the user used when registering their account. But other fields inside the `traits`
> section are supported as well.

Reviewer 3: ```suggestion
To start account recovery, Ory Identities (Ory Kratos) needs to know which address to send the recovery message to. In most cases this is the address the user provided when registering their account. Other fields inside the `traits` section are supported as well.
```

:::note

> If the email address used for recovery purposes is the same as the email used for verification purposes and the account hasn't

Reviewer 3: ```suggestion
If the email address used for recovery is the same as the email used for verification but the account hasn't
```

been activated yet (see [Account Activation](./verify-email-account-activation.mdx)), the address will also be verified by
completing the account recovery successfully.

:::

> To specify that a trait of the identity should be used for recovery, use the following identity schema:

Reviewer 3: ```suggestion
To specify a trait of the identity to be used for recovery, use the following identity schema:
```

             "credentials": {
               "password": {
                 "identifier": true
               }
             },
+            "recovery": {
+              "via": "email"
+            }
           }
         }
       }
       "additionalProperties": false
     }
   }
 }
```

## Email templates

<p>
  <figure>
    <img alt="Recovery email sent to unknown address" src={useBaseUrl("images/kratos/email-recovery-unknown.png")} />
    <figcaption>
      If the requested email address is an unknown recovery address, an account recovery attempt email is sent to that email
      address.
    </figcaption>
  </figure>
</p>

This prevents account enumeration attacks as explained in this
[blog post by Troy Hunt](https://www.troyhunt.com/website-enumeration-insanity-how-our-personal-data-is-leaked/).

<p>
  <figure>
    <img alt="Recovery email sent to a known recovery address" src={useBaseUrl("images/kratos/email-recovery-known.png")} />
    <figcaption>
      If the requested email address is a known recovery address, a recovery code is sent to that email address.
    </figcaption>
  </figure>
</p>

The emails are using templates that can be customized as explained in

> [Customizing E-Mail Templates](../../emails-sms/custom-email-templates).

Author: Sure, sounds good!

## Updating credentials

> After the user completed the recovery successfully they receive a privileged session and are redirected to the settings page. Here
> they can update their credentials.

Reviewer 3: ```suggestion
After the user completes the recovery successfully they receive a privileged session and are redirected to the settings page to update their credentials.
```

Read more about the [privileged sessions](../../session-management/01_overview.mdx#privileged-sessions).

## Invalidate other sessions

> In some scenarios it can be useful to deactivate all active sessions, once an account has been recovered. This forces anyone that
> currently has access to the account to re-authenticate.

Reviewer 3: ```suggestion
In some scenarios it can be useful to revoke all active sessions once an account has been recovered. This forces anyone with access to the account to re-authenticate.
```

```diff
 selfservice:
   flows:
     recovery:
       enabled: true
       ui_url: http://127.0.0.1:4455/recovery
+      after:
+        hooks:
+        - hook: revoke_active_sessions
```

## One-time passwords vs. magic links

> The one-time password method (`code`) is preferred, as magic links have multiple drawbacks:

Reviewer 3: ```suggestion
The one-time password method (`code`) is preferred as magic links have multiple drawbacks:
```

Author: The comma is correct here: https://www.magnumproofreading.com/post/blog-proper-punctuation-comma-before-as

- Some email virus scanners open links in emails to scan them. This invalidates the magic link and a user may not be able to
  recover their account.

> - API flows initialized by apps on mobile phones or smart devices are not possible when using magic links.

Reviewer: ```suggestion
- API flows initialized by apps on mobile phones or smart devices don't work with magic links.
```

> - Magic links often open in different browsers than the one used to initialize the flow, depending on the user's device settings.
>   This may result in a confusing user experience for your users.

Reviewer: ```suggestion
- Depending on the device settings, clicking a magic link can open a different browser than the one used to initialize the recovery flow. This can confuse your users. 
```

Ory Identities (Ory Kratos) uses one-time passwords by default. Magic links are available for legacy reasons and will be removed
in the future.

> :::important

Reviewer: ```suggestion
:::warning
```
Let's stick to general admonition types/names.

Author: I'd say this does not warrant the red "DANGER !!!" marker. And `:::important` is used quite a lot in our docs.

Author: Hm, I guess not "a lot". `:::important` is actually just undocumented, because it was removed...

But whatever, `:::caution` fits better here anyway. Thanks for the link!

> When updating an existing project from the `link` method to use the `code` method, there might be changes in the behavior of the
> UI. Double-check the flow after updating your configuration.

Reviewer: ```suggestion
When you change the recovery strategy from `link` to `code` in an existing project, you might need to adjust the UI to accommodate for that flow. Make sure the flow works correctly with your UI implementation after changing the recovery strategy.
```
Nice one, lemme one-up it ;)

Author: Lot of "flow"s in there...

I think there was some talk, that we want to remove the word "flow", as it's more of a technical term used internally in the codebase. So I've adjusted your proposal a bit more.

> UI. Double-check the flow, after updating your configuration.

Reviewer 3: ```suggestion
UI. Double-check the flow after updating your configuration.
```

:::

## Multifactor authentication

By design, the account recovery process bypasses the primary account credential. You can choose to require multifactor

> authentication for account recovery.

Author: The wording on the console (network) is not great, I've created an issue on our internal repo to invert the checkboxes and thus make the wording a bit more clear. 

However, this warning should stay in place IMO, as the account recovery effectively allows a user to bypass their primary account credential (e.g. the password) to get a privileged session so that the user can update their password. This is not great either, and we are in discussions on how to change that. But that is a little bit down the road.

>Judging by the UI text from the console, this setting affects all self-service flows?

That toggle only applies to the settings flow.

>If so, I think that this section should be turned into a reusable snippet that will be used across all ssf docs. I can take care of that if I deciphered the situation correctly ;)

Not sure what you mean by this, as this section only applies to this specific interaction between the recovery and settings flow.

> authentication for account recovery.

Reviewer: The setting we're talking about is in Authentication -> 2FA ->  General settings and by that I understand that this is a setting that affects all attempts of users trying to adjust their settings, not only those attempts triggered in the account recovery flow. Correct?

> authentication for account recovery.

Reviewer: If that's the case then I think it'll be much clearer if we mentioned that alongside the content from `## Updating credentials`. Happy to do that for you to avoid comment ping-pong.

> authentication for account recovery.

Author: Hmm, sure. I'll double check once you're done then. :)

> authentication for account recovery.

Author: >The setting we're talking about is in Authentication -> 2FA -> General settings and by that I understand that this is a setting that affects all attempts of users trying to adjust their settings, not only those attempts triggered in the account recovery flow. Correct?

Yes, that would apply to all "Settings Flows", regardless of whether the recovery is involved or not. We should add a disclaimer here, you're right.

> The account recovery process by design provides a direct bypass of the primary account credential. To ensure proper
> authentication, you can require a configured second factor to be provided.

Reviewer 3: ```suggestion
By design, the account recovery process bypasses the primary account credential. You can choose to require multifactor authentication for account recovery.
```

<Tabs
  defaultValue="oc"
  values={[
    {label: 'The Ory Network', value: 'oc'},
    {label: 'Self-hosted Ory Kratos config', value: 'oss'}
  ]}>
  <TabItem value="oc">
On the Ory Network Console go to <b>Two-Factor Authentication</b> and disable "Allow Self-Service Settings without Second Factor".
  </TabItem>
  <TabItem value="oss">

```yaml title="path/to/my/kratos/config.yml"
selfservice:
  flows:
    settings:
      required_aal: highest_available
```

  </TabItem>
</Tabs>

## Code examples for Node.js, React.js, Go

> The recovery user interface is a page in your own application (server, native app, single page app) that should render the actual

Reviewer: ```suggestion
The user interface for the account recovery and password reset flow is a page in your solution that should render the actual
```

Author: > account recovery and password reset flow

That's basically the same thing, so I'd remove "and password reset flow"

form elements for the user.

> In contrast to other identity systems, Ory Identities (Ory Kratos) doesn't render this HTML directly. Instead, you need to
> implement the HTML code in your application (for example in Node.js + Express.js, Java, PHP, React.js, ...), which gives you
> complete flexibility and customizability in your user interface flows and designs. This part of your application then directly
> interfaces with Ory Identities through an API.

Reviewer: ```suggestion
In contrast to other identity systems, Ory Identities (Ory Kratos) doesn't render this HTML directly. Instead, you need to
implement the HTML code in your solution, which gives you
complete flexibility and customizability in your user interface flows and designs. This part of your application then directly
interfaces with Ory Identities through the API.
```

Reviewer: Changed "an API" to "the API" - it's about Ory API, right?

Author: Yes, "the API" is also okay. 

We do have multiple APIs depending on the use-case, but in general we are refering to all Ory APIs here.

> In stark contrast to other identity systems, the Ory Identities (Ory Kratos) doesn't render this HTML directly. Instead, you need
> to implement the HTML code in your application (for example in Node.js + Express.js, Java, PHP, React.js, ...), which gives you
> extreme flexibility and customizability in your user interface flows and designs. This part of your application then directly
> interfaces with Ory Identities by calling different API endpoints.

Reviewer 3: ```suggestion
In contrast to other identity systems, the Ory Identities (Ory Kratos) doesn't render this HTML directly. Instead, you need
to implement the HTML code in your application (for example in Node.js + Express.js, Java, PHP, React.js, ...), which gives you
complete flexibility and customizability in your user interface flows and designs. This part of your application then directly
interfaces with Ory Identities through an API.
```

The API responds with a JSON document describing the form elements to render and actions the form should take upon submission,
cancellation, etc. The following shows examples for a few different languages and frameworks.

> <RenderFlow flow="recovery" />

Reviewer: On that topic: 
where is that coming from? Is that supposed to be the result of implementing the code that's presented in other tabs?

<img width="500" alt="image" src="https://user-images.githubusercontent.com/29116492/199027986-07b70004-eb26-44b8-b970-e4d5582a72d0.png">

> <RenderFlow flow="recovery" />

Reviewer: aaaand react is just 404

<img width="723" alt="image" src="https://user-images.githubusercontent.com/29116492/199028274-48b776e4-a69a-471f-900d-2a26db4c8616.png">

> <RenderFlow flow="recovery" />

Author: > Maybe that's not something that has to be necessary handled in this PR, but I really dislike this element (whatever you call it, I don't have the lingo at hand). Seems to be overcomplicated 

Element or Component :)

>On that topic: where is that coming from? Is that supposed to be the result of implementing the code that's presented in other tabs?

They're dynamic files included from our examples repositories. So basically example implementations of everything that is discussed on this page.

>aaaand react is just 404

That's the react native example, which does not work yet, as we need to implement some more stuff in kratos.
Due to the nature of the `<RenderFlow />` component, I don't think we can disable that view, though. ^^

The examples need some updating anyway, so I'll do that separately.

> <RenderFlow flow="recovery" />

Author: > Maybe that's not something that has to be necessary handled in this PR, but I really dislike this element (whatever you call it, I don't have the lingo at hand). Seems to be overcomplicated 

Element or Component :)

>On that topic: where is that coming from? Is that supposed to be the result of implementing the code that's presented in other tabs?

They're dynamic files included from our examples repositories. So basically example implementations of everything that is discussed on this page.

>aaaand react is just 404

That's the react native example, which does not work yet, as we need to implement some more stuff in kratos.
Due to the nature of the `<RenderFlow />` component, I don't think we can disable that view, though. ^^

The examples need some updating anyway, but I'll do that separately.

> <RenderFlow flow="recovery" />

Reviewer: > They're dynamic files included from our examples repositories. So basically example implementations of everything that is discussed on this page.

Sorry, I wasn't clear. What I meant is that the screenshot is just slapped there and there's no explanation why it's there. I'm guessing that it's supposed to be showing the UI you'll get by adding the code in other tabs to your app?

> <RenderFlow flow="recovery" />

Author: Yup, I guess. It's not my idea, I think we can improve that section a lot.

Screenshots in documentation mostly bad, as the UI changes, but the documentation is often not updated. I'd say we don't need the screenshot, but it's rendered via the `RenderFlow` element. 

Maybe we can do that in a different PR, let me know if you need input.

> <RenderFlow flow="recovery" />

Reviewer: Sure, no need to block this PR. I'd like to see `RenderFlow` gone soon though.

> These in turn respond with a defined JSON document describing the form elements to render and actions the form should take upon
> completion, cancellation, etc. The following shows examples for a few different languages and frameworks.

Reviewer 3: ```suggestion
The API responds with a JSON document describing the form elements to render and actions the form should take upon submission, cancellation, etc. The following shows examples for a few different languages and frameworks.
```

title: Recover access to accounts and perform password reset
sidebar_label: Account recovery
toc_max_heading_level: 4
---

import RenderFlow from "@site/src/theme/Code/RenderFlow"
import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"

Account recovery allows registered users to regain access to their account if they forget their password or lose access to the
second authentication factor required for multifactor authentication (MFA).

To regain access to their account, the user must prove that they are the owner of the account. The verification of ownership is
performed using the [recovery address defined by the user](#account-recovery-address). When the account recovery flow is
initiated, the system sends a link or a one-time code to the recovery address defined by the user. The user must access the link
or enter the code they received to confirm their identity and ownership of the account.

When the user accesses the configured recovery method, they receive a
[privileged session](../../session-management/overview#privileged-sessions) and are taken to their account's settings page where
they must create a new password to regain access to their account.

:::info

> The account recovery flow doesn't reset user's second authentication factor.

Author: This is a bit too vague, IMO, as it can absolutely be used to reset the second authentication factor. After all, the user is redirected to the settings page, that can be used to change the second factor.

What you probably mean is that, by default, the recovery flow does not require the second factor to be presented after entering the recovery code. The [part I wrote about that](https://github.com/ory/docs/pull/995/files/d4cd72fe889dd648a8a0ae307eb54edb018066ae..e453d5ed18c36812de26064ac6ed732af3152df5#diff-8d79b4961459ffe8dcfeb33b986c91b1bb486fb349c9dc6690f1edb49ef1f69eL205) is just gone without replacement here. What should we do about that?

Reviewer 2: Yes, please re-add that!

Reviewer: Why I removed that - enabling 2FA requirement for settings isn't something that's directly tied to the recovery flow. If you require 2FA, the user must provide their 2nd auth factor even if they're not doing the recovery. That's what I got from your comments/the UI text in the Console.

:::

## Supported recovery strategies

Ory Identities (Ory Kratos) supports two recovery strategies:

- Recovery by one-time passwords (`code`)
- Recovery by magic links (`link`)

### Comparison

> The "one-time password" strategy (`code`) is the default, preferred, and recommended by Ory.

Author: "Strategy" is a term internally used in Kratos. I think that term is quite abstract and might be hard to grasp for people wanting to implement Kratos in their application.

A better term is "Method" as used in the configuration as well, so we absolutely need to align the formulation here. For now, it's probably best to leave it like it is, but we should do a large scale refactoring to get rid of the term "Strategy" in favor of "Method". This is just my opinion, but I think Aeneas mentioned something like that before as well.

Reviewer: We can start with this one and switch to method, not a problem for me :)

  recovery flow. This can confuse your users.

:::caution

When you change the recovery strategy from `link` to `code` in an existing project, you might need to adjust your UI. Make sure
the flow works correctly with your UI implementation after changing the recovery strategy.

:::

## Configuration

````mdx-code-block
<Tabs groupId="oc-or-oss">
<TabItem value="oc" label="Ory Network" default>

To enable account recovery and adjust related options, go to **Authentication** → **Account Recovery** in the [Ory Console](https://console.ory.sh/guides)

</TabItem>
<TabItem value="oss" label="Self-hosted Ory Kratos config">

You can enable and configure account recovery in the Ory Kratos config:

```yaml title="path/to/my/kratos/config.yml"
selfservice:
methods:
 code: # Enables "one-time code" strategy. Use 'link' to enable "magic link" strategy.
   enabled: true
   config:
     lifespan: 15m # Defines how long the code or link are valid for. Default: 1h.

flows:
 recovery:
   enabled: true
   lifespan: 15m  # Defines how much time the user has to complete the recovery flow in the UI. Default: 1h.
   use: code # Defines which strategy is used, one of 'code' or 'link'.
```

> Both strategies require methods require sending emails to users. This means that you must configure an SMTP server for your instance.

Author: ```suggestion
Both strategies require sending emails to users. This means that you must configure an SMTP server for your instance.
```

 }
```

### Email templates

Ory Identities (Ory Kratos) comes with default email templates for recovery flows.

You can quickly replace the defaults and customize the messages to match the look and feel of your solution. Read

> [this document](../../emails-sms/custom-email-templates) to learn more.

Author: IMO, "this document" is not very descriptive and can be confusing if you're just skimming the page to find the link to the "custom email templates" page, for example.

I'd prefer the actual page name somewhere in there, but that's just my two cents from the perspective of an avid documentation reader. ;)

Reviewer: You really wouldn't skim the doc for this link specifically. This is just a comment that you can "learn more by reading something else". I prefer not using document titles in hyperlinks as it introduces wonkiness due to capitalization and makes the sentences unnaturally long. Also, I've been doing this across many docs that I've updated so this is also for consistency. Please keep it like this.

Author: Sure, this is just some general feedback, on stuff I noticed when trying to familiarize myself with Ory through the docs. We are, after all, writing this for developers and developers don't like reading a lot of text. So, the better "skimmable" a text is, the better, IMO. This is not blocking for me, resolving.
